src/main/java/org/jenkinsci/plugins/ssegateway/EventHistoryStore.java

```diff
@@ -0,0 +1,267 @@
+/*
+ * The MIT License
+ *
+ * Copyright (c) 2016, CloudBees, Inc.
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+package org.jenkinsci.plugins.ssegateway;
+
+import javax.annotation.CheckForNull;
+import javax.annotation.Nonnull;
+import java.io.File;
+import java.io.IOException;
+import java.nio.file.DirectoryStream;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.nio.file.Paths;
+import java.util.Map;
+import java.util.Timer;
+import java.util.TimerTask;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+
+import hudson.security.ACL;
+import org.apache.commons.io.FileUtils;
+import org.jenkins.pubsub.ChannelSubscriber;
+import org.jenkins.pubsub.Message;
+import org.jenkins.pubsub.PubsubBus;
+import org.kohsuke.accmod.Restricted;
+import org.kohsuke.accmod.restrictions.NoExternalUse;
+
+/**
+ * Channel event message history store.
+ * <p>
+ * Currently stores event history in files on disk, purging them
+ * as they go "stale" (after they expire).
+ * 
+ * @author <a href="mailto:tom.fennelly@gmail.com">tom.fennelly@gmail.com</a>
+ */
+@Restricted(NoExternalUse.class)
+public final class EventHistoryStore {
```

Comment:
I might be off here as I don't know if I understand uses of this class well, so bear with me. I think this class could have issues in terms of MT behavior. Its final and pretty much everything in the class is static. Some of the variables historyRoot and autoExpreTimer are mutable (not final), this means anytime its read/written must be locked. In fact most of the mutation methods must be synchronized. This might cause performance constraints assuming its going to be used by multiple threads. 

An alternative approach might be to keep class variable non-static and construct them using contractor and/or builder pattern to construct a non-mutable version of it - this can easy MT pressure and then you would need to synchronized only specific things based on use case.